sdk/src/azure/core/az_mqtt5_telemetry_consumer_hfsm.c

```diff
@@ -0,0 +1,398 @@
+ */
+static az_result ready(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  az_mqtt5_telemetry_consumer* this_policy = (az_mqtt5_telemetry_consumer*)me;
+
+  if (_az_LOG_SHOULD_WRITE(event.type))
+  {
+    _az_LOG_WRITE(event.type, AZ_SPAN_FROM_STR("az_telemetry_consumer/ready"));
+  }
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+    case AZ_HFSM_EVENT_EXIT:
+      // No-op
+      break;
+
+    case AZ_MQTT5_EVENT_SUBACK_RSP:
+    {
+      // if get suback that matches the sub we sent, stop waiting for the suback
+      az_mqtt5_suback_data* data = (az_mqtt5_suback_data*)event.data;
+      if (data->id == this_policy->_internal.pending_subscription_id)
+      {
+        _subscribe_stop_timer(this_policy);
+        this_policy->_internal.pending_subscription_id = 0;
+      }
+      // else, keep waiting for the proper suback
+      break;
+    }
+
+    case AZ_HFSM_EVENT_TIMEOUT:
+    {
+      if (event.data == &this_policy->_internal.telemetry_consumer_timer)
+      {
+        // if subscribing times out, go to faulted state - this is not recoverable
+        _az_RETURN_IF_FAILED(_az_hfsm_transition_peer((_az_hfsm*)me, ready, faulted));
+      }
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_PUB_RECV_IND:
+    {
+      az_mqtt5_recv_data* recv_data = (az_mqtt5_recv_data*)event.data;
+      az_mqtt5_telemetry_consumer_codec_data telemetry_event_data;
+
+      // Check if the publish is on the telemetry topic
+      if (az_result_succeeded(az_mqtt5_telemetry_consumer_codec_parse_received_topic(
+              this_policy->_internal.telemetry_consumer_codec,
+              recv_data->topic,
+              &telemetry_event_data)))
+      {
+        // clear subscription timer if we get a pub on the topic, since that implies we're
+        // subscribed
+        if (this_policy->_internal.pending_subscription_id != 0)
+        {
+          _subscribe_stop_timer(this_policy);
+          this_policy->_internal.pending_subscription_id = 0;
+        }
+
+        // parse the telemetry details and send it to the application
+        if (az_result_failed(_handle_telemetry(this_policy, recv_data)))
+        {
+          if (_az_LOG_SHOULD_WRITE(AZ_HFSM_EVENT_ERROR))
+          {
+            _az_LOG_WRITE(
+                AZ_HFSM_EVENT_ERROR,
+                AZ_SPAN_FROM_STR("az_telemetry_consumer/ready Error handling incoming publish - "
+                                 "missing required content type property"));
+          }
+        }
+      }
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_PUBACK_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_OPEN_REQ:
+    case AZ_MQTT5_EVENT_CONNECT_RSP:
+      break;
+
+    default:
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+/**
+ * @brief Failure state - locks up all execution of this hfsm
+ */
+static az_result faulted(az_event_policy* me, az_event event)
```

Comment:
To follow the convention in the `az_mqtt5_connection_hfsm`s faulted state it would be good to have this state only return `AZ_ERROR_HFSM_INVALID_STATE` and have the api callback on the transition back in line 217

Reply:
Left a comment about this on your PR as well - the entry state should succeed to indicate that faulted was entered correctly and then any other events after that should fail. Let's sync on this once you test in the connection_hfsm